PyRDF/backend/AWS.py

```diff
@@ -92,20 +131,78 @@ def invoke_root_lambda(client, root_range, script):
                 'filelist': str(root_range.filelist),
                 'friend_info': encode_object(root_range.friend_info)
             })
-            return client.invoke(
-                FunctionName='root_lambda',
-                InvocationType='Event',
-                Payload=bytes(payload, encoding='utf8')
-            )
+
+            # Maybe here give info about number of invoked lambda for awsmonitor
+            #self.logger.info(f'New lambda - 31')
+
+            while trials:
+                trials -= 1
+                try:
+                    #my_before_time = time.time()
+                    response = client.invoke(
+                        FunctionName='root_lambda',
+                        InvocationType= 'Event', #'RequestResponse',
+                        Payload=bytes(payload, encoding='utf8')
+                    )
+                    #my_after_time = time.time()
+                    #self.logger.info(f'Invocation time: {my_after_time - my_before_time}')
+
+                    if 'FunctionError' in response:
+                        raise RuntimeError('Lambda runtime error')
+
+                    response = process_response(response)
+
+                    if 'Payload' in response:
+                        response_payload = response['Payload']
+                        lambda_status = response_payload['statusCode'] if 'statusCode' in response_payload else 200
+                        if lambda_status != 200:
+                            raise RuntimeError(f'Lambda status code {lambda_status}')
+                except botocore.exceptions.ClientError as error:
+                    # AWS site errors
+                    self.logger.warn(error['Error']['Message'])
+                except RuntimeError as error:
+                    # Lambda runtime errors and lambdas that
+                    # returned statusCode different than 200
+                    # (maybe also response processing errors)
+                    self.logger.warn(error)
+                except Exception as error:
+                    # All other errors
+                    self.logger.warn(error)
+                else:
```

Comment:
is this else even needed at this point?

Reply:
Code under this else is executed only if there is no exception, what indicates that the loop should be ended straight away, cause lambda invocation have been successful. Idk, can change it.